🎤 feat: add custom speech config, browser TTS/STT features, and dynamic speech tab settings

berry-13 opened the pull request:
## Summary

This PR introduces several key features and improvements related to the speech functionality in LibreChat:

- **Custom Speech Configuration**: Added a custom speech configuration option in `librechat.yaml`, allowing the ADMIN to set pre-configured "speech tab" settings
- **Browser TTS Language Selection**: Enabled language selection for text-to-speech (TTS) directly within the browser
- **Browser STT Streaming**: Implemented streaming for speech-to-text (STT) in the browser
- **Dynamic Speech Tab Settings**: The speech tab settings now dynamically appear and disappear based on user settings (e.g., browser/external engine dropdown)
- **Refactoring**:
  - Renamed `endpointSTT` and `endpointTTS` to `engineSTT` and `engineTTS` respectively
  - Moved the speech API to a subpath `/api/files/speech`

### Breaking Changes

- The variables `SpeechToText` and `TextToSpeech` in the store have been renamed to `speechToText` and `textToSpeech`. If you encounter any issues, please delete LibreChat's cache
- The speech section has been moved under the `speech:`

## Change Type

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Testing

### External STT/TTS:

- **ElevenLabs**
- **OpenAI**

### Local STT/TTS:

- **Chrome**

### Test Configuration:

To reproduce the test process, follow these steps:

1. Configure the `librechat.yaml` file with appropriate speech settings
2. Select a TTS language in the browser and verify the speech output
3. Test STT streaming functionality in the browser using various input sources
4. Verify that the speech tab settings appear or disappear based on the selected user settings
5. Ensure all refactored endpoints (`engineSTT` and `engineTTS`) function correctly
6. Confirm the speech API is accessible via the new subpath `/api/files/speech`

## Checklist

- [x] My code adheres to this project's style guidelines.
- [x] I have performed a self-review of my own code.
- [x] I have commented in any complex areas of my code.
- [x] I have made pertinent documentation changes.
- [x] My changes do not introduce new warnings.
- [x] I have written tests demonstrating that my changes are effective or that my feature works.
- [x] Local unit tests pass with my changes.
- [x] Any changes dependent on mine have been merged and published in downstream modules.
- [x] A pull request for updating the documentation has been submitted.

berry-13 committed:
feat: update useTextToSpeech and useSpeechToText hooks to support external audio endpoints

This commit updates the useTextToSpeech and useSpeechToText hooks in the Input directory to support external audio endpoints. It introduces the useGetExternalTextToSpeech and useGetExternalSpeechToText hooks, which determine whether the audio endpoints should be set to 'browser' or 'external' based on the value of the endpointTTS and endpointSTT Recoil states. The useTextToSpeech and useSpeechToText hooks now use these new hooks to determine whether to use external audio endpoints

berry-13 committed:
feat: add userSelect style to ConversationModeSwitch label

berry-13 committed:
fix: remove unused updateTokenWebsocket function and import

The updateTokenWebsocket function and its import are no longer used in the OpenAIClient module. This commit removes the function and import to clean up the codebase

berry-13 committed:
feat: support external audio endpoints in useTextToSpeech and useSpeechToText hooks

This commit updates the useTextToSpeech and useSpeechToText hooks in the Input directory to support external audio endpoints. It introduces the useGetExternalTextToSpeech and useGetExternalSpeechToText hooks, which determine whether the audio endpoints should be set to 'browser' or 'external' based on the value of the endpointTTS and endpointSTT Recoil states. The useTextToSpeech and useSpeechToText hooks now use these new hooks to determine whether to use external audio endpoints

berry-13 committed:
feat: update AutomaticPlayback component to AutomaticPlaybackSwitch; tests: added AutomaticPlaybackSwitch.spec
>
> This commit renames the AutomaticPlayback component to AutomaticPlaybackSwitch in the Speech directory. The new name better reflects the purpose of the component and aligns with the naming convention used in the codebase.

berry-13 committed:
feat: update useSpeechToText hook to include interimTranscript

This commit updates the useSpeechToText hook in the client/src/components/Chat/Input/AudioRecorder.tsx file to include the interimTranscript state. This allows for real-time display of the speech-to-text transcription while the user is still speaking. The interimTranscript is now used to update the text area value during recording.

berry-13 committed:
feat: Add customConfigSpeech API endpoint for retrieving custom speech configuration

This commit adds a new API endpoint  in the  file under the  directory. This endpoint is responsible for retrieving the custom speech configuration using the  function from the  module

berry-13 committed:
feat: update store var  and ; fix: getCustomConfigSpeech

danny-avila commented on api/server/routes/files/index.js:
```diff
@@ -20,8 +19,7 @@ const initialize = async () => {
 
-  router.use('/stt', stt);
-  router.use('/tts', tts);
+  router.use('/speech', speech);
```
update your branch and you'll see an important note on route initialization order

danny-avila commented on client/src/hooks/Input/useSpeechToTextBrowser.ts:
```diff
@@ -1,22 +1,44 @@
 
   const toggleListening = () => {
     if (browserSupportsSpeechRecognition) {
+      if (!isMicrophoneAvailable) {
```
`if (!browserSupportsSpeechRecognition) {`

early return instead of nesting

danny-avila commented on client/src/hooks/Input/useSpeechToTextBrowser.ts:
```diff
@@ -28,7 +50,7 @@ const useSpeechToTextBrowser = () => {
   useEffect(() => {
     const handleKeyDown = (e: KeyboardEvent) => {
-      if (e.shiftKey && e.altKey && e.code === 'KeyL' && endpointSTT === 'browser') {
+      if (e.shiftKey && e.altKey && e.code === 'KeyL' && !useExternalSpeechToText) {
```
is `useExternalSpeechToText` a hook or a boolean value? you should not name a variable starting with `use` in react unless it's a hook

danny-avila commented on client/src/hooks/Input/useSpeechToTextBrowser.ts:
```diff
@@ -37,15 +59,19 @@ const useSpeechToTextBrowser = () => {
+
   return {
-    isListening: listening,
+    isListening: isListening,
```
redundant

berry-13 committed:
fix: client tests, removed unused import

berry-13 committed:
feat: Update useCustomConfigSpeechQuery to return an array of custom speech configurations

This commit modifies the useCustomConfigSpeechQuery function in the client/src/data-provider/queries.ts file to return an array of custom speech configurations instead of a single object. This change allows for better handling and manipulation of the data in the application

berry-13 committed:
feat: Update useCustomConfigSpeechQuery to return an array of custom speech configurations

berry-13 committed:
refactor: Update variable name in speechTab schema

berry-13 committed:
refactor: removed unused and nested code

berry-13 committed:
fix: using recoilState

berry-13 committed:
refactor: Update Speech component to use useCallback for setting settings

berry-13 committed:
Merge branch 'main' into speech-refactor

danny-avila commented:
please fix the broken test

berry-13 committed:
Merge branch 'speech-refactor' of https://github.com/danny-avila/librechat into speech-refactor

berry-13 committed:
fix: test

berry-13 committed:
fix: tests

danny-avila commented:
slated for v0.7.4

berry-13 committed:
Merge branch 'main' into speech-refactor

berry-13 committed:
Merge branch 'main' into speech-refactor

danny-avila committed:
Merge branch 'main' into speech-refactor

danny-avila commented:
Hi Berry, there are elevated errors when using this without config settings

```yaml
memory, and will not scale past a single process.
2024-06-21 09:31:55 info: Server listening on all interfaces at port 3080. Use http://localhost:3080 to access it
2024-06-21 09:31:59 error: Failed to get speechTab settings: Configuration or speechTab schema is missing
2024-06-21 09:31:59 error: Failed to get voices: Configuration or TTS schema is missing
2024-06-21 09:32:45 error: Failed to get speechTab settings: Configuration or speechTab schema is missing
2024-06-21 09:32:45 error: Failed to get voices: Configuration or TTS schema is missing
```

I should not be seeing any errors when I don't have speech enabled/configured and I open the Settings

danny-avila commented:
Also when i add the new format for speech settings, i should not get an error if i don't have speechTab settings
```bash
2024-06-21 09:36:42 error: Failed to get speechTab settings: Configuration or speechTab schema is missing
```

danny-avila commented:
lastly, if I'm using the old setup, this error message is confusing and I will think the app has a bug:
```bash
2024-06-21 09:40:49 error: Invalid custom config file at /home/danny/LibreChat/librechat.yaml [
  {
    "code": "unrecognized_keys",
    "keys": [
      "tts",
      "stt"
    ],
    "path": [],
    "message": "Unrecognized key(s) in object: 'tts', 'stt'"
  }
]
```

this is a parsing error for the custom config file, we should still show it, but also add a note if this error message is detected, that the format has changed.

Also please accompany this PR with an update to the changelog: https://www.librechat.ai/changelog

berry-13 committed:
Merge branch 'main' into speech-refactor

berry-13 committed:
feature: ensure that the settings don't change after modifying then through the UI

berry-13 committed:
remove comment

berry-13 committed:
Merge branch 'main' into speech-refactor

berry-13 committed:
fix: Handle error gracefully in getCustomConfigSpeech and getVoices endpoints

berry-13 committed:
fix: Handle error

berry-13 committed:
fix: backend tests

danny-avila committed:
Merge branch 'main' into speech-refactor

danny-avila committed:
fix: invalid custom config logging

danny-avila commented:
> lastly, if I'm using the old setup, this error message is confusing and I will think the app has a bug:
> 
> ```shell
> 2024-06-21 09:40:49 error: Invalid custom config file at /home/danny/LibreChat/librechat.yaml [
>   {
>     "code": "unrecognized_keys",
>     "keys": [
>       "tts",
>       "stt"
>     ],
>     "path": [],
>     "message": "Unrecognized key(s) in object: 'tts', 'stt'"
>   }
> ]
> ```
> 
> this is a parsing error for the custom config file, we should still show it, but also add a note if this error message is detected, that the format has changed.
> 
> Also please accompany this PR with an update to the changelog: https://www.librechat.ai/changelog

@berry-13 this is what I meant: https://github.com/danny-avila/LibreChat/pull/2921/commits/51c5d0488a155b773e8692efcc630569371f7dfe

danny-avila committed:
chore: add back custom config info logging

danny-avila committed:
chore: revert loadCustomConfig spec

berry-13 commented:
> @berry-13 this is what I meant: [51c5d04](https://github.com/danny-avila/LibreChat/commit/51c5d0488a155b773e8692efcc630569371f7dfe)

oh ok, sorry. I was thinking about every key, not just `stt` and `tts`

dkindlund commented:
Hey @danny-avila and @berry-13 -- I was trying to get the `whisper-1` model to work with `stt` using the `LiteLLM` proxy.

My `librechat.yaml` looks like:
```
speech:
  stt:
    openai:
      url: 'https://litellmproxy/v1/audio/transcriptions'
      apiKey: 'sk-1234'
      model: 'whisper-1'
 ```
 
 When I try using external STT, I get an error in the client and these are the messages I'm seeing in the server logs:
 ```
 An error occurred while preparing the OpenAI API STT request:  Cannot read properties of undefined (reading 'openai')
 An error occurred while processing the audio: Cannot read properties of undefined (reading 'status')
 ```
 
 Is there something else I'm missing to make STT work in LibreChat?  Thanks!
